File: src/components/context/editors/instance/keys.ts
/**
 * Composite-key contract for instance disable lists.
 *
 * Spec invariant: composite keys are JSON-encoded string arrays, no
 * whitespace, unicode literal (not escaped), array order preserved as
 * caller-given. Python `engine.modules._keys.encode_key` MUST output
 * identical bytes for identical input. Cross-language fitness test in
 * `tests/test_engine_keys.py` enforces this.
 *
 * See: docs/superpowers/specs/2026-05-07-instance-overrides-modal-design.md §4.5
 */

export function encodeKey(parts: readonly string[]): string {
  return JSON.stringify(parts);
}

export function decodeKey(key: string): string[] {
  try {
    const parsed = JSON.parse(key);
    return Array.isArray(parsed) ? parsed.map(String) : [];
  } catch {
    return [];
  }
}

File: engine/modules/_keys.py
"""Composite-key encoder. Mirrors JS `encodeKey` byte-for-byte.

Spec invariant: composite keys are JSON-encoded string arrays, no
whitespace, unicode literal (not escaped), array order preserved.
Cross-language fitness test in `tests/test_engine_keys.py` enforces
byte-exact match.

See: docs/superpowers/specs/2026-05-07-instance-overrides-modal-design.md §4.5
"""
import json


def encode_key(parts: list[str]) -> str:
    return json.dumps(parts, separators=(",", ":"), ensure_ascii=False)


def decode_key(key: str) -> list[str]:
    """Inverse of encode_key. Round-trip safe by spec invariant.

    Raises ValueError on malformed input — caller should treat unknown
    keys as "skip silently" rather than crash the resolve.
    """
    parts = json.loads(key)
    if not isinstance(parts, list) or not all(isinstance(p, str) for p in parts):
        raise ValueError(f"composite key must decode to list[str], got {parts!r}")
    return parts
